feat: document the server-side generation of Delta Artifacts (#1967)
mendersoftware/mender-docs: 06.Artifact-creation/05.Server-Side-Generation-of-Delta-Artifacts/docs.md

---
title: Server-Side Generation of Delta Artifacts
taxonomy:
    category: docs
    label: tutorial
---

Mender Enterprise supports the Server-Side generation of Delta Artifacts.
This document explains the prerequisites to enable this feature and how it works.

!!! Hosted Mender supports this feature in the Enterprise plan; however, its availability requires a tenant-specific activation.
!!! If you are a customer interested in it, please get in touch with the [Hosted Mender support team](mailto:support@mender.io).

### Prerequisites

To support the Server-Side Generation of Delta Artifacts, your device must report to the server the availability of the `mender-binary-delta` Update Module.
This will happen automatically if you are building your firmware using Yocto and the latest version of the `meta-mender` layer.

> Otherwise, please make sure your device is reporting the following inventory keys to the Mender Server:
> 
> * `rootfs-image.update-module.mender-binary-delta.mender_update_module: mender-binary-delta`
> * `rootfs-image.update-module.mender-binary-delta.version: <version>`

Reviewer: Inventory keys won't be sufficient, these need to be in the `mender show-provides` (no inventory is sent to `deployments/next` endpoint, only the provides).

The alternative should be building and installing a empty artifact, with something like:
```suggestion
Otherwise, create and install your own bootstrap Artifact to provide these keys:

\```
mender-artifact write bootstrap-artifact \
        --artifact-name mender-binary-delta-provide-keys \
        --device-type <dev-type> \
        --provides "rootfs-image.update-module.mender-binary-delta.mender_update_module:mender-binary-delta" \
        --provides "rootfs-image.update-module.mender-binary-delta.version:<version>" \
        --output-path mender-binary-delta-provide-keys.mender

The Artifact can be installed either in standalone mode or deployed remotely with the Mender server.
\```
```

> Additionally, to make the Server-Side Generation of Delta Artifacts possible, both the Mender Artifacts of the current version of the firmware (`rootfs`) running on your device and the updated one must be available as Artifacts on the Mender Server.

Reviewer: The delta Artifact (as per today) is done from only one Artifact:
```suggestion
Additionally, to make the Server-Side Generation of Delta Artifacts possible, both the Mender Artifact of the current version of the firmware (`rootfs`) running on your device and the updated one must be available as Artifacts on the Mender Server.
```

Reviewer: Also "of the current version of the firmware ...", rephrase to "of the current firmware version"

Reviewer 2: > The delta Artifact (as per today) is done from only one Artifact

not sure I understand; you need two artifacts to generate delta

Author: I agree, we need both; it's plural because you need both the current version AND the updated one.

Reviewer: It might be my grammar miss-understanding. But let me retry,

To me in reads like "both _the_ Mender Artifacts ... running on your device (meaning: all the Artifacts running there) and _the_ updated one...".

"I like both _the_ cake and _the_ ice cream" is not the same as "I like both _the_ cakes and _the_ ice cream"

Author: I reworded it, and I dropped "both" and made Mender Artifact singular.

### Enabling the Server-Side generation of Delta Artifacts

You can enable this feature by selecting the option "Generate and deploy Delta Artifacts (where available)" when you create a deployment for your devices. This option will enable the Server-Side generation of Delta Artifacts when turned on.

![Mender logo](deployment.jpg)

When you target a device running the Mender Client and the Binary Delta Update Module (see Prerequisites above for details) with a deployment where this option is active, the Mender Server tries to look up the correct Delta Artifact from the firmware (`rootfs`) version currently running on the Device to the desired version you selected with the deployment.

If the Delta Artifact already exists, the Mender Server selects it and sends the corresponding deployment instructions to the Device. However, if the Delta Artifact does not exist, it will start a Delta Artifact generation job in the background to create it. In this case, the Device will temporarily receive a `No updates` response for the Server. Once the generation job completes and the Device asks for updates again, the Server will serve the generated Delta Artifact to the Device.

> Please note that if the Server cannot generate the Delta Artifact, it will automatically fall back to the non-Delta Artifact. This can happen, for example, because the Device is not reporting the availability of the Binary Delta Update Module through the inventory, or if the Mender Artifact containing the firmware (`rootfs`) version the Device is currently running on is not available in the Mender Server.

Reviewer: ```suggestion
Please note that if the Server cannot generate the Delta Artifact, it will automatically fall back to the non-Delta Artifact. This can happen, for example, because the Device is not reporting the availability of the Binary Delta Update Module through the `provides` database keys, or if the Mender Artifact containing the firmware (`rootfs`) version the Device is currently running on is not available in the Mender Server.
```

### Configuration settings

You can customize the Delta Artifact generation settings in the Global settings of your tenant.

> You can specify a timeout for the generation job; the Mender Server will abort it if it takes more than the given amount of seconds.

Reviewer 2: not entirely true, because currently it's not possible to abort the job, but at least we can serve full image after timeout

---
title: Server-Side Generation of Delta Artifacts
taxonomy:
    category: docs
    label: tutorial
---

Mender Enterprise supports the Server-Side generation of Delta Artifacts.
This document explains the prerequisites to enable this feature and how it works.

!!! Hosted Mender supports this feature in the Enterprise plan; however, its availability requires a tenant-specific activation.
!!! If you are a customer interested in it, please get in touch with the [Hosted Mender support team](mailto:support@mender.io).

### Prerequisites

> To support the Server-Side Generation of Delta Artifacts, your device must report to the server the availability of the `mender-binary-delta` Update Module.

Reviewer 3: Why is "Server-Side Generation" always capitalized? To me this is neither a term that refers to our brand, nor is it mentioned in the [taxonomy](https://docs.mender.io/overview/taxonomy), so my feeling is that is should be common casing in the English language.

> This will happen automatically if you are building your firmware using Yocto and the latest version of the `meta-mender` layer.

Reviewer 3: Not only on the latest, that might be confusing.

    --provides "rootfs-image.update-module.mender-binary-delta.version:<version>" \
    --output-path mender-binary-delta-provide-keys.mender
```

The Artifact can be installed in standalone mode or remotely deployed with the Mender server.

Additionally, to make the Server-Side Generation of Delta Artifacts possible, both the Mender Artifacts of the current version of the firmware (`rootfs`) running on your device and the updated one must be available as Artifacts on the Mender Server.

### Enabling the Server-Side generation of Delta Artifacts

You can enable this feature by selecting the option "Generate and deploy Delta Artifacts (where available)" when you create a deployment for your devices. This option will enable the Server-Side generation of Delta Artifacts when turned on.

![Mender logo](deployment.jpg)

When you target a device running the Mender Client and the Binary Delta Update Module (see Prerequisites above for details) with a deployment where this option is active, the Mender Server tries to look up the correct Delta Artifact from the firmware (`rootfs`) version currently running on the Device to the desired version you selected with the deployment.

> If the Delta Artifact already exists, the Mender Server selects it and sends the corresponding deployment instructions to the Device. However, if the Delta Artifact does not exist, it will start a Delta Artifact generation job in the background to create it. In this case, the Device will temporarily receive a `No updates` response for the Server. Once the generation job completes and the Device asks for updates again, the Server will serve the generated Delta Artifact to the Device.

Reviewer 3: ```suggestion
If the Delta Artifact already exists, the Mender Server selects it and sends the corresponding deployment instructions to the Device. However, if the Delta Artifact does not exist, it will start a Delta Artifact generation job in the background to create it. In this case, the Device will temporarily receive a `No updates` response for the Server. In the next update polling cycle after the Delta Artifact generation has finished, the resulting Delta artifact will be served to the Device.
```

> Please note that if the Server cannot generate the Delta Artifact, it will automatically fall back to the non-Delta Artifact. This can happen, for example, because the Device is not reporting the availability of the Binary Delta Update Module through the `provides` database keys, or if the Mender Artifact containing the firmware (`rootfs`) version the Device is currently running on is not available in the Mender Server.

Reviewer 3: ```suggestion
Please note that if the Server cannot generate the Delta Artifact, it will automatically fall back to the non-Delta Artifact. Possible reasons are:
- the Device is not reporting the availability of the Binary Delta Update Module through the `provides` database keys
- the Mender Artifact containing the firmware (`rootfs`) version the Device is currently running on is not available in the Mender Server.
```

### Configuration settings

You can customize the Delta Artifact generation settings in the Global settings of your tenant.
You can specify a timeout for the generation job; the Mender Server will serve the full image if the Server-Side generation of the Delta Artifact takes more than the given amount of seconds.
You can also set the xDelta3 arguments to customize the Delta generation and the compression level.
Please refer to the [Create a Delta update Artifact](../05.Create-a-Delta-update-Artifact/) chapter for more information about the available options.

![Mender logo](settings.jpg)

> The settings apply to the newly generated Delta Artifacts only. Existing Delta Artifacts previously updated to the Mender Server won't be affected by the changes to these values.

Reviewer 3: ```suggestion
The settings apply to the newly generated Delta Artifacts only. Existing Delta Artifacts previously uploaded to the Mender Server won't be affected by the changes to these values.
```
